File: forge/src/main/java/me/pandamods/pandalib/neoforge/platform/NetworkHelperImpl.java
/*
 * Copyright (C) 2024 Oliver Froberg (The Panda Oliver)
 *
 * This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  any later version.
 *
 * You should have received a copy of the GNU Lesser General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package me.pandamods.pandalib.neoforge.platform;

import dev.architectury.utils.Env;
import me.pandamods.pandalib.event.events.NetworkingEvents;
import me.pandamods.pandalib.networking.NetworkContext;
import me.pandamods.pandalib.networking.NetworkReceiver;
import me.pandamods.pandalib.platform.Services;
import me.pandamods.pandalib.platform.services.NetworkHelper;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.server.level.ServerPlayer;
import net.neoforged.neoforge.network.PacketDistributor;
import net.neoforged.neoforge.network.event.RegisterPayloadHandlersEvent;
import net.neoforged.neoforge.network.handling.DirectionalPayloadHandler;
import net.neoforged.neoforge.network.registration.PayloadRegistrar;

public class NetworkHelperImpl extends PayloadRegistrar implements NetworkHelper {
	public NetworkHelperImpl() {
		super("1");
	}

	@Override
	public <T extends CustomPacketPayload> void registerClientReceiver(CustomPacketPayload.Type<T> type,
																	   StreamCodec<? super RegistryFriendlyByteBuf, T> codec,
																	   NetworkReceiver<T> receiver) {
		this.playToClient(type, codec, (arg, ctx) -> receiver.receive(new NetworkContext(ctx.player(), Env.CLIENT), arg));
	}

	@Override
	public <T extends CustomPacketPayload> void registerServerReceiver(CustomPacketPayload.Type<T> type,
																	   StreamCodec<? super RegistryFriendlyByteBuf, T> codec,
																	   NetworkReceiver<T> receiver) {
		this.playToServer(type, codec, (arg, ctx) -> receiver.receive(new NetworkContext(ctx.player(), Env.SERVER), arg));
	}

	@Override
	public <T extends CustomPacketPayload> void registerBiDirectionalReceiver(CustomPacketPayload.Type<T> type,
																			  StreamCodec<? super RegistryFriendlyByteBuf, T> codec,
																			  NetworkReceiver<T> clientReceiver,
																			  NetworkReceiver<T> serverReceiver) {
		this.playBidirectional(type, codec, new DirectionalPayloadHandler<>(
				(arg, ctx) -> clientReceiver.receive(new NetworkContext(ctx.player(), Env.CLIENT), arg),
				(arg, ctx) -> serverReceiver.receive(new NetworkContext(ctx.player(), Env.SERVER), arg)
		));
	}

	public static void registerPackets(final RegisterPayloadHandlersEvent event) {
		NetworkingEvents.PACKET_PAYLOAD_REGISTRY.invoker().register(Services.NETWORK);
	}

	@Override
	public <T extends CustomPacketPayload> void sendToServer(T payload) {
		PacketDistributor.sendToServer(payload);
	}

	@Override
	public <T extends CustomPacketPayload> void sendToPlayer(ServerPlayer player, T payload) {
		PacketDistributor.sendToPlayer(player, payload);
	}

	@Override
	public <T extends CustomPacketPayload> void sendToAllPlayers(T payload) {
		PacketDistributor.sendToAllPlayers(payload);
	}
}

File: neoforge/src/main/kotlin/me/pandamods/pandalib/neoforge/platform/NetworkHelperImpl.kt
/*
 * Copyright (C) 2024 Oliver Froberg (The Panda Oliver)
 *
 * This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  any later version.
 *
 * You should have received a copy of the GNU Lesser General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
package me.pandamods.pandalib.neoforge.platform

import dev.architectury.utils.Env
import me.pandamods.pandalib.event.events.NetworkingEvents
import me.pandamods.pandalib.networking.NetworkContext
import me.pandamods.pandalib.networking.NetworkReceiver
import me.pandamods.pandalib.platform.Services
import me.pandamods.pandalib.platform.services.NetworkHelper
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.server.level.ServerPlayer
import net.neoforged.neoforge.network.PacketDistributor
import net.neoforged.neoforge.network.event.RegisterPayloadHandlersEvent
import net.neoforged.neoforge.network.handling.DirectionalPayloadHandler
import net.neoforged.neoforge.network.handling.IPayloadContext
import net.neoforged.neoforge.network.registration.PayloadRegistrar

class NetworkHelperImpl : PayloadRegistrar("1"), NetworkHelper {
	override fun <T : CustomPacketPayload> registerClientReceiver(
		type: CustomPacketPayload.Type<T>,
		codec: StreamCodec<in RegistryFriendlyByteBuf, T>,
		receiver: NetworkReceiver<T>
	) {
		playToClient<T>(type, codec) { arg: T, ctx: IPayloadContext -> receiver.receive(NetworkContext(ctx.player(), Env.CLIENT), arg) }
	}

	override fun <T : CustomPacketPayload> registerServerReceiver(
		type: CustomPacketPayload.Type<T>,
		codec: StreamCodec<in RegistryFriendlyByteBuf, T>,
		receiver: NetworkReceiver<T>
	) {
		playToServer<T>(type, codec) { arg: T, ctx: IPayloadContext -> receiver.receive(NetworkContext(ctx.player(), Env.SERVER), arg) }
	}

	override fun <T : CustomPacketPayload> registerBiDirectionalReceiver(
		type: CustomPacketPayload.Type<T>,
		codec: StreamCodec<in RegistryFriendlyByteBuf, T>,
		clientReceiver: NetworkReceiver<T>,
		serverReceiver: NetworkReceiver<T>
	) {
		playBidirectional<T>(
			type, codec,
			DirectionalPayloadHandler<T>(
				{ arg: T, ctx: IPayloadContext -> clientReceiver.receive(NetworkContext(ctx.player(), Env.CLIENT), arg) },
				{ arg: T, ctx: IPayloadContext -> serverReceiver.receive(NetworkContext(ctx.player(), Env.SERVER), arg) }
			)
		)
	}

	override fun <T : CustomPacketPayload> sendToServer(payload: T) {
		PacketDistributor.sendToServer(payload)
	}

	override fun <T : CustomPacketPayload> sendToPlayer(player: ServerPlayer, payload: T) {
		PacketDistributor.sendToPlayer(player, payload)
	}

	override fun <T : CustomPacketPayload> sendToAllPlayers(payload: T) {
		PacketDistributor.sendToAllPlayers(payload)
	}

	companion object {
		fun registerPackets(event: RegisterPayloadHandlersEvent) {
			NetworkingEvents.PACKET_PAYLOAD_REGISTRY.invoker().register(Services.NETWORK)
		}
	}
}
